### ISSUE TITLE
Several comments about API design

### ISSUE THREAD

**Issue by zhujiangang** [microsoft/DeepSpeed#704]
1. There are too many irrelevant pieces in the DeepSpeedEngine, etc. lr scheduler, gradient accumulation, tensorboard, checkpoint, training details, data loader, amp, fp16, etc. The core of DeepSpeedEngine is zero optimizer, right? It should be as clean as something like DDP in torch, amp in Apex, by providing several simple wrappers over models and optimizers or an engine (context) to maintain some distributed computing setting.
2. Why there is a timer in the stage 2 optimizer but not in stage 1 optimizer? More general, why need to force user to use these timer or logging mechanism. We should leave it to users.
3. Why not to expose another API to initialize the engine from code instead of from a json file? 
4. Seems that 'dist_init_required' is not used in engine.py

**Comment by tjruwase** [comment]
@zhujiangang Thanks for your feedback. I will try to address them below.
1) Although ZeRO is an important part of DeepSpeed engine, it is not the core of the library. The goal of DeepSpeed is to help model scientists in large-scale model training. And from our experience many of those pieces, lr scheduler, gradient accumulation, checkpoint, etc. are incredibly useful and of varying importance to the wide diversity of model scientists out there. 

2) The objective of these timers is not to force users to use them. We provide [configuration knobs](https://www.deepspeed.ai/docs/config-json/#logging) to control the timer and logging. If you encounter issues with these controls, please help us by open issues against them. Users are always free to use their own logging and timer mechanisms in just like the [Megatron example](https://github.com/microsoft/DeepSpeedExamples/blob/1bee84f6eb75ed7e39e34601bfdd66d79cafe99a/Megatron-LM/pretrain_gpt2.py#L255).

3) We do have an API to initialize engine with a `dict` instead of json fille as [here](https://deepspeed.readthedocs.io/en/latest/_modules/deepspeed.html#initialize).

4) I believe `dist_init_required `is used in engine.py, such as [here](https://github.com/microsoft/DeepSpeed/blob/248f6383a1426390a33981793a22fe3dec78b855/deepspeed/runtime/engine.py#L130-L134).
